dex/runner.go

```diff
@@ -108,15 +108,26 @@ func (c *ConnectionMaster) Connect(ctx context.Context) (err error) {
 	return err
 }
 
-// Disconnect closes the connection and waits for shutdown.
+// Disconnect closes the connection and waits for shutdown. This is safe to use
+// on an unconnected ConnectionMaster as long as it was constructed with
+// NewConnectionMaster.
 func (c *ConnectionMaster) Disconnect() {
-	c.cancel()
 	c.mtx.RLock()
+	defer c.mtx.RUnlock()
```

Comment:
AFAICT, there's no contention on the `ConnectionMaster`'s (`contextManager`'s) mutex.  In fact, I'm not even sure it is needed in a proper use pattern, but there may be different uncoordinated goroutines that can each do `Connect` and `Disconnect` so I'm not just nuking it.

Reply:
I think calling `Connect` twice is veering into the territory of improper use of the type's methods, but eliminating the mutex would be lovely.  Just like doing `Run(ctx)` more than once (at least with the first call still running) for just about all `Runners` is an implicitly prohibited pattern that applies here too I think.  I suppose we need to consider the design goals of `ConnectionMaster`, but I can see it going either way - single use or allow repeated connect/disconnect calls.

Ideally all of `ConnectionMaster` would be not have any synchronization mechanisms and the consumer would just be aware of its state... I'm starting to recall why I kept this in draft now.  Hmm, I might revert most of the last two commits and go back to the `if oldWallet.connected()` check since I'm clearly touching on a design issue with `ConnectionMaster` (and how `Core` manages them) that I don't know I want to solve in this bugfix PR.